File: 7. Reverse Integer/7. Reverse Integer.py
class Solution:
    def reverse(self, x: int) -> int:
        INT_MAX : int = 2**31 - 1
        INT_MIN : int = -2**31

        res : int = 0
        sign : int = -1 if x < 0 else 1
        x = abs(x)

        while x != 0:
            digit : int = x % 10
            x //= 10

            # Check overflow before multiplying
            if res > (INT_MAX - digit) // 10:
                return 0
            
            res : int = res * 10 + digit
        
        return sign * res

File: 7. Reverse Integer/7. Reverse Integer.js
/**
 * @param {number} x
 * @return {number}
 */
var reverse = function(x) {
    const INT_MAX = 2 ** 31 - 1
    const INT_MIN = Math.pow(-2, 31)

    let res = 0
    const sign = x < 0 ? -1 : 1
    x = Math.abs(x)

    while (x != 0){
        let digit = x % 10
        x = Math.floor(x / 10)

        // Check overflow before multiplying
        if (res > (INT_MAX - digit) / 10){
            return 0
        }

        res = res * 10 + digit
    }

    return sign * res
};